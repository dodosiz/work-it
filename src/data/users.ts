import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AppState } from "./store";
import { v4 } from "uuid";

export interface UsersState {
	users: User[];
	userFormOpened: boolean;
}

export interface User {
	id: string;
	firstName: string;
	lastName: string;
	role: string;
}

interface AddUserPayload {
	firstName: string;
	lastName: string;
	role: string;
}

interface DeleteUserPayload {
	id: string;
}

const initialState: UsersState = {
	users: [],
	userFormOpened: false,
};

const usersSlice = createSlice({
	name: "users",
	initialState,
	reducers: {
		addUser: (state, action: PayloadAction<AddUserPayload>) => {
			state.users.push({
				id: v4(),
				...action.payload,
			});
		},
		deleteUser: (state, action: PayloadAction<DeleteUserPayload>) => {
			const index = state.users.findIndex(
				(user) => user.id === action.payload.id
			);
			state.users.splice(index, 1);
		},
		openUserForm: (state) => {
			state.userFormOpened = true;
		},
		closeUserForm: (state) => {
			state.userFormOpened = false;
		},
	},
});

export const { addUser, deleteUser, openUserForm, closeUserForm } =
	usersSlice.actions;
export const usersReducer = usersSlice.reducer;

export const usersSelector = (state: AppState) => state.usersState.users;
export const userFormOpenedSelector = (state: AppState) =>
	state.usersState.userFormOpened;
