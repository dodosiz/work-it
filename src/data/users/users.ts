import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AppState } from "../store";
import { v4 } from "uuid";

export interface UsersState {
	users: User[];
	userIdToEdit: string | undefined;
	userFormOpened: boolean;
}

export interface User {
	id: string;
	firstName: string;
	lastName: string;
	role: string;
}

interface AddUserPayload {
	firstName: string;
	lastName: string;
	role: string;
}

interface UpdateUserPayload {
	id: string;
	firstName: string;
	lastName: string;
	role: string;
}

interface UserIdPayload {
	id: string;
}

const initialState: UsersState = {
	users: [],
	userIdToEdit: undefined,
	userFormOpened: false,
};

const usersSlice = createSlice({
	name: "users",
	initialState,
	reducers: {
		addUser: (state, action: PayloadAction<AddUserPayload>) => {
			state.users.push({
				id: v4(),
				...action.payload,
			});
		},
		deleteUser: (state, action: PayloadAction<UserIdPayload>) => {
			const index = state.users.findIndex(
				(user) => user.id === action.payload.id
			);
			state.users.splice(index, 1);
		},
		editUser: (state, action: PayloadAction<UserIdPayload>) => {
			state.userIdToEdit = action.payload.id;
		},
		updateUser: (state, action: PayloadAction<UpdateUserPayload>) => {
			const index = state.users.findIndex(
				(user) => user.id === action.payload.id
			);
			state.users.splice(index, 1, action.payload);
		},
		openUserForm: (state) => {
			state.userFormOpened = true;
		},
		closeUserForm: (state) => {
			state.userFormOpened = false;
		},
	},
});

export const {
	addUser,
	deleteUser,
	editUser,
	updateUser,
	openUserForm,
	closeUserForm,
} = usersSlice.actions;
export const usersReducer = usersSlice.reducer;

export const usersSelector = (state: AppState) => state.usersState.users;
export const userFormOpenedSelector = (state: AppState) =>
	state.usersState.userFormOpened;
export const editedUserSelector = (state: AppState) => {
	return state.usersState.users.find(
		(u) => u.id === state.usersState.userIdToEdit
	);
};
