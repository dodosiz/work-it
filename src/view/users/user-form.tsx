import * as React from "react";
import { Button, Form } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { addNotification } from "../../data/notifications/notifications";
import {
	addUser,
	closeUserForm,
	editedUserSelector,
	updateUser,
} from "../../data/users/users";

export function UserForm() {
	const dispatch = useDispatch();
	const editedUser = useSelector(editedUserSelector);
	const [firstName, setFirstName] = React.useState(
		editedUser ? editedUser.firstName : ""
	);
	const [lastName, setLastName] = React.useState(
		editedUser ? editedUser.lastName : ""
	);
	const [role, setRole] = React.useState(editedUser ? editedUser.role : "");
	const handleFirstNameChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setFirstName(event.target.value);
	};
	const handleLastNameChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setLastName(event.target.value);
	};
	const handleRoleChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setRole(event.target.value);
	};
	const handleSubmit: React.FormEventHandler<HTMLFormElement> = (event) => {
		event.preventDefault();
		if (editedUser) {
			dispatch(updateUser({ id: editedUser.id, firstName, lastName, role }));
			dispatch(
				addNotification({
					message: `Updated user ${firstName} ${lastName}.`,
				})
			);
		} else {
			dispatch(addUser({ firstName, lastName, role }));
			dispatch(
				addNotification({
					message: `Added user ${firstName} ${lastName}.`,
				})
			);
		}
		dispatch(closeUserForm());
	};
	const handleCancel = () => {
		dispatch(closeUserForm());
	};
	const submitDisabled = () => {
		return !(firstName.length && lastName.length && role.length);
	};
	return (
		<Form onSubmit={handleSubmit}>
			<Form.Group className="mb-3" controlId="firstName">
				<Form.Label>First name:</Form.Label>
				<Form.Control
					type="text"
					value={firstName}
					onChange={handleFirstNameChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="lastName">
				<Form.Label>Last name:</Form.Label>
				<Form.Control
					type="text"
					value={lastName}
					onChange={handleLastNameChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="role">
				<Form.Label>Role:</Form.Label>
				<Form.Control type="text" value={role} onChange={handleRoleChange} />
			</Form.Group>
			<Button disabled={submitDisabled()} variant="primary" type="submit">
				Save
			</Button>{" "}
			<Button variant="secondary" onClick={() => handleCancel()}>
				Cancel
			</Button>
		</Form>
	);
}
