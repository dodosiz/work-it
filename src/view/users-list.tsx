import * as React from "react";
import { useSelector } from "react-redux";
import { User, usersSelector } from "../data/users";

export function UsersList() {
	const users = useSelector(usersSelector);
	return <ul>
		{users.map((user, index) => (
			<User key={`$${user.name}-${index}`} user={user}/>
		))}
	</ul>;
}

interface UserProps {
	user: User;
}

function User(props: UserProps) {
	return <li>{props.user.name}, role: {props.user.role}</li>;
}
