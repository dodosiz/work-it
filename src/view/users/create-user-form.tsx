import * as React from "react";
import { Button, Form } from "react-bootstrap";
import { useDispatch } from "react-redux";
import { addUser, closeUserForm } from "../../data/users";

export function CreateUserForm() {
	const dispatch = useDispatch();
	const [firstName, setFirstName] = React.useState("");
	const [lastName, setLastName] = React.useState("");
	const [role, setRole] = React.useState("");
	const handleFirstNameChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setFirstName(event.target.value);
	};
	const handleLastNameChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setLastName(event.target.value);
	};
	const handleRoleChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setRole(event.target.value);
	};
	const handleSubmit: React.FormEventHandler<HTMLFormElement> = (event) => {
		event.preventDefault();
		dispatch(addUser({ firstName, lastName, role }));
		dispatch(closeUserForm());
	};
	const submitDisabled = () => {
		return !(firstName.length && lastName.length && role.length);
	};
	return (
		<Form onSubmit={handleSubmit}>
			<Form.Group className="mb-3" controlId="firstName">
				<Form.Label>First name:</Form.Label>
				<Form.Control
					type="text"
					value={firstName}
					onChange={handleFirstNameChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="lastName">
				<Form.Label>Last name:</Form.Label>
				<Form.Control
					type="text"
					value={lastName}
					onChange={handleLastNameChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="role">
				<Form.Label>Role:</Form.Label>
				<Form.Control type="text" value={role} onChange={handleRoleChange} />
			</Form.Group>
			<Button disabled={submitDisabled()} variant="primary" type="submit">
				Submit
			</Button>
		</Form>
	);
}
