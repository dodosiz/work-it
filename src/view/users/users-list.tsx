import * as React from "react";
import { Table, Button } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { deleteUser, User, usersSelector } from "../../data/users";
import { BsFillTrashFill } from "react-icons/bs";

export function UsersList() {
	const dispatch = useDispatch();
	const users = useSelector(usersSelector);
	const handleDelete = (id: string) => {
		dispatch(deleteUser({ id }));
	};
	return (
		<Table striped bordered hover>
			<thead>
				<tr>
					<th>#</th>
					<th>First Name</th>
					<th>Last Name</th>
					<th>Role</th>
					<th>Action</th>
				</tr>
			</thead>
			<tbody>
				{users.map((user, index) => (
					<User
						key={user.id}
						user={user}
						index={index}
						handleDelete={handleDelete}
					/>
				))}
			</tbody>
		</Table>
	);
}

interface UserProps {
	index: number;
	user: User;
	handleDelete(id: string): void;
}

function User(props: UserProps) {
	return (
		<tr>
			<td>{props.index}</td>
			<td>{props.user.firstName}</td>
			<td>{props.user.lastName}</td>
			<td>{props.user.role}</td>
			<td>
				<Button
					variant="danger"
					onClick={() => props.handleDelete(props.user.id)}
				>
					<BsFillTrashFill />
				</Button>
			</td>
		</tr>
	);
}
