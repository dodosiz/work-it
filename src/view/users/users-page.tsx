import * as React from "react";
import { useDispatch, useSelector } from "react-redux";
import {
	closeUserForm,
	editedUserSelector,
	userFormOpenedSelector,
} from "../../data/users/users";
import { ModalDialog } from "../modal/modal-dialog";
import { UserForm } from "./user-form";
import { UsersList } from "./users-list";

export function UsersPage() {
	const dispatch = useDispatch();
	const userFormOpened = useSelector(userFormOpenedSelector);
	const editedUser = useSelector(editedUserSelector);
	const handleCloseUserForm = () => {
		dispatch(closeUserForm());
	};
	return (
		<div>
			<ModalDialog
				handleClose={handleCloseUserForm}
				show={userFormOpened}
				title={editedUser ? "Edit user" : "Create new user"}
			>
				<UserForm />
			</ModalDialog>
			<UsersList />
		</div>
	);
}
