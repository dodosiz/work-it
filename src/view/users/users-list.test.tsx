import { configureStore, EnhancedStore } from "@reduxjs/toolkit";
import * as React from "react";
import { Provider } from "react-redux";
import { usersReducer } from "../../data/users/users";
import { UsersList } from "./users-list";
import { render, RenderResult, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("Users list component:", () => {
	let store: EnhancedStore;
	let component: RenderResult;
	beforeEach(() => {
		store = configureStore({
			reducer: {
				usersState: usersReducer,
			},
			preloadedState: {
				usersState: {
					users: [
						{
							id: "1",
							firstName: "George",
							lastName: "Daniels",
							role: "developer",
						},
						{
							id: "2",
							firstName: "Nick",
							lastName: "Gates",
							role: "graphic designer",
						},
					],
					userFormOpened: false,
					userIdToEdit: undefined,
				},
			},
		});
		component = render(
			<Provider store={store}>
				<UsersList />
			</Provider>
		);
	});
	test("snapshot test", () => {
		expect(component.container).toMatchSnapshot();
	});
	test("delete a user", () => {
		userEvent.click(screen.getByTestId("delete-button-1"));
		expect(store.getState().usersState.users).toEqual([
			{
				id: "1",
				firstName: "George",
				lastName: "Daniels",
				role: "developer",
			},
		]);
	});
	test("edit a user", () => {
		userEvent.click(screen.getByTestId("edit-button-1"));
		expect(store.getState().usersState.userIdToEdit).toEqual("2");
	});
});
