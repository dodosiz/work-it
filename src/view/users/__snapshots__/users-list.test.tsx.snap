// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users list component: displays a list of users 1`] = `
<div>
  <table
    class="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          First Name
        </th>
        <th>
          Last Name
        </th>
        <th>
          Role
        </th>
        <th>
          Action
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          0
        </td>
        <td>
          George
        </td>
        <td>
          Daniels
        </td>
        <td>
          developer
        </td>
        <td>
          <button
            class="btn btn-danger"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 1 0z"
              />
            </svg>
          </button>
        </td>
      </tr>
      <tr>
        <td>
          1
        </td>
        <td>
          Nick
        </td>
        <td>
          Gates
        </td>
        <td>
          graphic designer
        </td>
        <td>
          <button
            class="btn btn-danger"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 1 0z"
              />
            </svg>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
